components/apps_form/apps_form_field/apps_form_field.tsx

```diff
@@ -0,0 +1,206 @@
+        } else if (field.options) {
+            const options = field.options.map(({label, value}) => ({text: label, value}));
+            this.provider = new MenuActionProvider(options);
+        }
+
+        return this.provider;
+    }
+
+    render() {
+        const {
+            field,
+            name,
+            value,
+            onChange,
+            errorText,
+            listComponent,
+        } = this.props;
+
+        const placeholder = field.hint || '';
+
+        const displayName = (field.modal_label || field.label) as string;
+        let displayNameContent: React.ReactNode = (field.modal_label || field.label) as string;
+        if (field.is_required) {
+            displayNameContent = (
+                <React.Fragment>
+                    {displayName}
+                    <span className='error-text'>{' *'}</span>
+                </React.Fragment>
+            );
+        }
+
+        const helpText = field.description;
+        let helpTextContent: React.ReactNode = helpText;
+        if (errorText) {
+            helpTextContent = (
+                <React.Fragment>
+                    {helpText}
+                    <div className='error-text mt-3'>
+                        {errorText}
+                    </div>
+                </React.Fragment>
+            );
+        }
+
+        if (field.type === 'text') {
+            const subtype = field.subtype || 'text';
+
+            let maxLength = field.max_length;
+            if (!maxLength) {
+                if (subtype === 'textarea') {
+                    maxLength = TEXTAREA_DEFAULT_MAX_LENGTH;
+                } else {
+                    maxLength = TEXT_DEFAULT_MAX_LENGTH;
+                }
+            }
+
+            let textType: InputTypes = 'input';
+            if (subtype && TextSetting.validTypes.includes(subtype)) {
+                textType = subtype as InputTypes;
+            }
+
+            const textValue = value as string;
+            return (
+                <TextSetting
+                    autoFocus={this.props.autoFocus}
+                    id={name}
+                    disabled={field.readonly}
+                    type={textType}
+                    label={displayNameContent}
+                    maxLength={maxLength}
+                    value={textValue || ''}
+                    placeholder={placeholder}
+                    helpText={helpTextContent}
+                    onChange={onChange}
+                    resizable={false}
+                />
+            );
+        } else if (field.type === 'channel' || field.type === 'user') {
+            let selectedValue: string | undefined;
+            if (this.props.value) {
+                selectedValue = (this.props.value as AppSelectOption).label;
+            }
+            return (
+                <AutocompleteSelector
+                    id={name}
+                    disabled={field.readonly}
+                    providers={[this.getProvider()]}
```

Comment:
I've changed it so they are stored on the component itself. The reasons why I've chosen not to put it into state are:

- There is currently no state in the component
- The only time we would want to update the providers is if we are switching to a channel or user field from a different type, with the same field name. This is a very niche use case, and 2/5 it's not worth supporting with the complications of `componentDidUpdate` / storing state for `getDerivedStateFromProps`.
- The interactive dialog fields actually store the providers on the component directly as well, because the component is stateless